In [1]:
#File and Library Importing
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\mojos\OneDrive\Career Foundry Course\Task4 - InstaCart Project\02 Data'

In [3]:
df_prods = pd.read_csv(os.path.join(path,'Raw Data','products.csv'),index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path,'Prepared Data','orders_wrangled.csv'),index_col = False)

In [5]:
#Mixed data training
df_test = pd.DataFrame()

In [6]:
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Checking for Mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird]) > 0:
        print(col)

mix


C:\Users\mojos\AppData\Local\Temp\ipykernel_25056\2046225207.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
#Checking for Missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [12]:
#subsetting the missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
#Checking and removing duplicates
df_prods_clean.shape

(49677, 5)

In [16]:
#Creating a subset of dupes for future review (not actually executed)
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups.shape

(5, 5)

In [18]:
#Creating the actual cleaned Dataframe w/ additional check for new size
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data','products_checked.csv'))

# Task 4.5 

In [21]:
#Step 2. Overview of df_ords

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#Nothing untoward about the data:
Unnamed: Essentially an additional index, not necessarily required

Order_id: Also nothing to note, this the descriptive stats aren't too important since the it's more of a string than a numeric value that will have calculations on it. Min/Max match row count values. (see shape below)

User_id: Mostly the same comments as order_id, but max value does not mirror row count due to the fact that users use the same ID when placing subsequent orders

Order_number: Represents the order count relating to a particular user - Lowest number is 1 (which makes sense as they would have to log atleast 1 order to be in the system, and max is 100 - a higher number given the context of what we are dealing with - but not outlandish)

Order_day_of_week: all whole numbers between 0-6 (expected 7 days) - count matches total rows
Order_hour_of_day: all whole numbers between 0-23 (expected 34 hour period) - count matches total rows

Days_since_prior_order: values themselves are whole numbers between 0-30 - no issues, 0 could represent 2 orders within a day (missed item from cart perhaps?) - total count does not represent the same count as dataframe rows BUT this could be due to NaN values which are indicative of a client's first order.

In [22]:
df_ords.shape

(3421083, 8)

In [23]:
#Step 3/4.
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_ords[weird])>0:
        print(col)

C:\Users\mojos\AppData\Local\Temp\ipykernel_25056\4150001944.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\mojos\AppData\Local\Temp\ipykernel_25056\4150001944.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\mojos\AppData\Local\Temp\ipykernel_25056\4150001944.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\mojos\AppData\Local\Temp\ipykernel_25056\4150001944.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\mojos\AppData\Local\Temp\ip

In [24]:
#No mixed data types in the output

In [25]:
# Step 5. Missing Values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### As per the description given during the decribe comment: 
Days_since_prior_order: total count does not represent the same count as dataframe rows BUT this could be due to NaN values which are indicative of a client's first order.
#### This is related to 206209 missing values, which as expected, is also the max number for the user_id field (indicating how many unique cliens there)

In [26]:
#Step 6. Dealing with Missing Values
# Initial impression would be to keep these but either a) add a flag column, or b) subset the missing values and label it something orbvious like df_client_first_order
# A last option could be to impute a 0 as the days since prior orders - but this would impact mean and median value calculations in the future.

#Option b.
df_client_first_orders = df_ords[df_ords['days_since_prior_order'].isnull()==True]


In [27]:
df_client_first_orders.shape

(206209, 8)

In [28]:
#Option A. Trying to create a flag variable [Assistance given using CF forum]
df_ords['first_order'] = np.where(np.isnan(df_ords['days_since_prior_order'].values), 'First Order','Repeat Customer')

In [29]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order      first_order  
0                  8                     NaN      First Order  
1                  7                    15.0  Repeat Customer  
2                 12                    21.0  Repeat Customer  
3                  7                    29.0  Repeat Customer  
4                 15                    28.0  Repeat Customer

In [30]:
df_ords['first_order'] = df_ords['first_order'].astype('category')

In [31]:
df_ords['first_order'].dtype

CategoricalDtype(categories=['First Order', 'Repeat Customer'], ordered=False, categories_dtype=object)

In [37]:
df_ords['first_order'].value_counts(dropna=False)

first_order
Repeat Customer    3214874
First Order         206209
Name: count, dtype: int64

##### I've generated a flag in a new column on the data. This is because the first order placed on an account is a vital piece of information regarding the customers lifecycle with instacart.

In [32]:
#Step 7.
df_ords_dupes = df_ords[df_ords.duplicated()]

In [33]:
df_ords_dupes

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

In [34]:
#No duplicates found within the data set, no actions required. If actions were required we would need to use the df.drop_duplicates() command.

In [35]:
#Step 9. Exporting the updated Order commands

df_ords.to_csv(os.path.join(path,'Prepared Data','orders_checked.csv'))

In [36]:
df_client_first_orders.to_csv(os.path.join(path,'Prepared Data','first_time_orders.csv'))